### PR TITLE
Transform Extension Collider and Bounds Calculation Performance Update

### PR DESCRIPTION
**Pull request by StephenHodgson**
# XRTK - Mixed Reality Toolkit Change Request

## Overview

I was having performance issues in my app when using the bounding box and manipulation handlers. I was able to narrow this down to a bounds calculation we were calling in each update loop.

See https://github.com/XRTK/SDK/pull/130

## Target of the change:

Is this enhancement for:

- Extension

## Changes:

- updated `Transfrom.GetColliderBounds` to reference the colliders it's using to get the bounds
- Added another `Transform.SetCollidersActive` override extension method that also uses a reference to the colliders used

## Breaking Change

- `Transfrom.GetColliderBounds` now requires you pass in a `ref` to the colliders you want to calculate the bounds for.

### CODE REVIEW

File: XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs [hunk 1]
```diff
@@ -104,7 +104,7 @@ private static IEnumerable<Transform> EnumerateHierarchyCore(this Transform root
         /// <param name="colliders">Optional cached collider collection.</param>
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs]:
```suggestion
        /// <param name="colliders">Referenced Collider collection that will be filled with data if none passed.</param>
```

Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs]: @tomzorz better?

File: XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs [hunk 2]
```diff
@@ -101,10 +101,10 @@ private static IEnumerable<Transform> EnumerateHierarchyCore(this Transform root
         /// <param name="syncTransform">
         /// True, by default, this will sync the <see cref="transform"/> rotation to calculate the axis aligned orientation.
         /// </param>
-        /// <param name="colliders">Optional cached collider collection.</param>
+        /// <param name="colliders">Referenced Collider collection that will be filled with data if none passed.</param>
         /// <returns>The total bounds of all colliders attached to this GameObject.
         /// If no colliders attached, returns a bounds of center and extents 0</returns>
-        public static Bounds GetColliderBounds(this Transform transform, bool syncTransform = true, Collider[] colliders = null)
+        public static Bounds GetColliderBounds(this Transform transform, ref Collider[] colliders, bool syncTransform = true)
```
Comment (FejZa) [inline XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs]: Documentation is missing for `syncTransform`

File: XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs [hunk 3]
```diff
@@ -338,8 +338,9 @@ public static bool TryGetDepth(Transform target, Transform parent, ref int depth
         /// <param name="transform"></param>
         /// <param name="direction"></param>
         /// <param name="bounds"></param>
+        /// <param name="cachedColliders"></param>
         /// <returns></returns>
-        public static Vector3 GetPointOnBoundsEdge(this Transform transform, Vector3 direction, Bounds bounds = default)
+        public static Vector3 GetPointOnBoundsEdge(this Transform transform, Vector3 direction, Bounds bounds = default, Collider[] cachedColliders = null)
```
Comment (FejZa) [inline XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs]: All input parameters documentation is missing.

File: XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs [hunk 4]
```diff
@@ -417,6 +418,25 @@ public static Collider[] SetCollidersActive(this Transform transform, bool isAct
             return colliders;
         }
 
+        /// <summary>
+        /// Sets the collider and all child colliders active with the provided value.
+        /// </summary>
+        /// <param name="transform"></param>
+        /// <param name="isActive"></param>
+        /// <param name="colliders"></param>
+        public static void SetCollidersActive(this Transform transform, bool isActive, ref Collider[] colliders)
```
Comment (FejZa) [inline XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs]: All input parameters documentation is missing.

**Comment by tomzorz** [comment]
I added the requested documentation but apparently I don't have push access to this branch, so I put it up here in a gist https://gist.github.com/tomzorz/8463f41ce14a6613bbc3dce95ffabde1

@FejZa or @StephenHodgson you can commit it :)

**Comment by StephenHodgson** [comment]
@tomzorz you are part of the xrtk org. you should have push writes.

You can also suggest the edits inline in the browser interface as well

**Comment by tomzorz** [comment]
> @tomzorz you are part of the xrtk org. you should have push writes.
> 
> You can also suggest the edits inline in the browser interface as well

¯\_(ツ)_/¯ I'm getting 403 when trying to push.

**Comment by tomzorz** [comment]
@StephenHodgson but please don't make me edit these in with the github UI, just take the file :)

**Comment by StephenHodgson** [comment]
You could always branch and open a PR against this one ¯\_(ツ)_/¯

**Comment by tomzorz** [comment]
@StephenHodgson no I can't, something is not right with the permissions.

image

File: XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs (hunk 2 above)
Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs]: it is there.

File: XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs (hunk 3 above)
Comment (StephenHodgson) [reply XRTK-Core/Packages/com.xrtk.core/Extensions/TransformExtensions.cs]: likely just needs to be copy pasted from above

**Pull request merged**